File: src/1335-Minimum-Difficulty-of-a-Job-Schedule/1335.cpp
class Solution 
{
public:
    int minDifficulty(vector<int>& jobDifficulty, int d) 
    {
        int n = jobDifficulty.size();
        if (n < d) return -1;
        
        int dp[n + 1] = {0};
        for (int i = n - 1; i >= 0; i--) dp[i] = max(dp[i + 1], jobDifficulty[i]);
        
        for (int t = 1; t <= d; t++)
        {
            for (int i = 0; i <= n - t; i++)
            {
                int m = 0; dp[i] = 10010;
                for (int j = i; j <= n - t; j++)
                {
                    m = max(m, jobDifficulty[j]);
                    dp[i] = min(dp[i], m + dp[j + 1]);
                }
            }
        }
        return dp[0];
    }
};

File: src/1335-Minimum-Difficulty-of-a-Job-Schedule/1335.py
from functools import lru_cache
class Solution:
    def minDifficulty(self, jobDifficulty: List[int], d: int) -> int:
        n = len(jobDifficulty)
        if n < d:
            return -1
        
        @lru_cache(None)
        def dfs(u, d):
            if d == 1:
                return max(jobDifficulty[u:])
            
            res, m = float("inf"), 0
            for i in range(u, n - d + 1):
                m = max(m, jobDifficulty[i])
                res = min(res, dfs(i + 1, d - 1) + m)
            return res
        
        return dfs(0, d)